### ISSUE TITLE
Yet another crash on launch, with call stack provided.

### ISSUE THREAD

**Issue by nizmow** [microsoft/terminal#4646]
# Environment

```none
Windows build number: 10.0.18363.0
Windows Terminal version (if applicable): Master branch

Any other software?
```

# Steps to reproduce

Yet another crash on startup issue! (I've checked all the others I can find but nothing seems applicable).

# Expected behavior

Launches!

# Actual behavior

Crashes after briefly showing a blank window.

Details:

I'm building from source as company AD lockdown prevents me from running the application any other way. Previously, building from the 0.8 tag everything was fine. Upgrading to 0.9, I get a crash on launch. I moved to master to check for a resolution, but I get the same behaviour. Since I've build from source I can attach the call stack.

![image](https://user-images.githubusercontent.com/176703/74839519-b1cb2f80-5325-11ea-89d6-fc6185c21588.png)

I've completely cleared my profiles and data directories, but it didn't make a difference. I have Powershell 6, Powershell 7 RC and WSL installed.

**Comment by DHowett-MSFT** [comment]
Thanks for the call stack. None of us are seeing this, so I really appreciate it.

Do you, per chance, have a profile that exits immediately on startup? It seems like Terminal is trying to set focus to a tab that doesn't fully exist.

**Comment by onekopaka** [comment]
I was seeing a crash with a very similar call stack, the startup profile I had just launched PowerShell 6, and PowerShell was able to launch fine. I disabled using the title of the active tab as the title of the window, and I was able to get it launched fine.

Unfortunately I don't have a crash dump handy, but I definitely recall the _OnTabSelectionChanged function, and it leading down to an Access Violation in winrt, in UnconditionalReleaseRef.

It hasn't been happening recently on commit 941a44a46491e1a018419eb24c9d92d3b2640e55 (I just haven't updated recently) however. I am on Windows build 18363.657.

**Comment by nizmow** [comment]
My profiles are all defaults, as I removed my existing profiles.json and let Terminal generate a new one. I can't see any reason why any of the configured profiles would do that. I've tried to change the default profile to various others and I get the same result. Default profiles.json that's generated is as follows:

``` json
// To view the default settings, hold "alt" while clicking on the "Settings" button.
// For documentation on these settings, see: https://aka.ms/terminal-documentation

{
    "$schema": "https://aka.ms/terminal-profiles-schema",

    "defaultProfile": "{61c54bbd-c2c6-5271-96e7-009a87ff44bf}",

    "profiles":
    {
        "defaults":
        {
            // Put settings here that you want to apply to all profiles
        },
        "list":
        [
            {
                // Make changes here to the powershell.exe profile
                "guid": "{61c54bbd-c2c6-5271-96e7-009a87ff44bf}",
                "name": "Windows PowerShell",
                "commandline": "powershell.exe",
                "hidden": false
            },
            {
                // Make changes here to the cmd.exe profile
                "guid": "{0caa0dad-35be-5f56-a8ff-afceeeaa6101}",
                "name": "cmd",
                "commandline": "cmd.exe",
                "hidden": false
            },
            {
                "guid": "{574e775e-4f2a-5b96-ac1e-a2962a402336}",
                "hidden": false,
                "name": "PowerShell",
                "source": "Windows.Terminal.PowershellCore"
            },
            {
                "guid": "{049b895c-b0b8-55de-bce3-27d3220d7ab3}",
                "hidden": false,
                "name": "PowerShell Core 6",
                "source": "Windows.Terminal.PowershellCore"
            },
            {
                "guid": "{c6eaf9f4-32a7-5fdc-b5cf-066e8a4b1e40}",
                "hidden": false,
                "name": "Ubuntu-18.04",
                "source": "Windows.Terminal.Wsl"
            }
        ]
    },

    // Add custom color schemes to this array
    "schemes": [],

    // Add any keybinding overrides to this array.
    // To unbind a default keybinding, set the command to "unbound"
    "keybindings": []
}
```

**Comment by shunf4** [comment]
**tl;dr: update Visual Studio.**

---

Same problem here. I investigated into it, finding that if I attempt to iterate through vector `_tabs` using this code: 
```
for (auto tab : _tabs)
{
}
```
after [one tab is appended to it](https://github.com/microsoft/terminal/blob/4f8acb4b9fbd61196d0a5287749a58abe0fbaa28/src/cascadia/TerminalApp/TerminalPage.cpp#L493), then the exception happens.
That's because, in the `GetAt()` function (generated code), `winrt_impl_result` is somehow not properly initialized:
```
template <typename D, typename T> WINRT_IMPL_AUTO(T) consume_Windows_Foundation_Collections_IVector<D, T>::GetAt(uint32_t index) const
{
    T winrt_impl_result{ empty_value<T>() };  // not properly initialized
    check_hresult(WINRT_IMPL_SHIM(Windows::Foundation::Collections::IVector<T>)->GetAt(index, put_abi(winrt_impl_result)));
    return winrt_impl_result;
}
```
IIRC even if I set breakpoints inside `empty_value` they aren't hit. Hence, `(*((winrt::Windows::Foundation::IUnknown*)&(*((winrt::Windows::Foundation::IInspectable*)&(*((winrt::TerminalApp::ITab*)&winrt_impl_result)))))).m_ptr` equals `0xcccccccccccccccc` after initialization when the `put_api` function executes. In `put_abi`, this line of code executes:
```
object = nullptr;
```
where `object` is `winrt_impl_result`. This tries to release the reference to the invalid address `0xcccccccccccccccc`, then the whole program crashed.

My solution is to update Visual Studio to the latest version. After that, the code recomples and program seems to work sucessfully.

**Comment by DHowett-MSFT** [comment]
So, there was an unfortunate issue in VS 16.4.3 that impacted iterating over WinRT IVectors (or similar enumerables) in C++/WinRT. 16.4.4 contained the fix for that one.

**Comment by DHowett-MSFT** [comment]
@nizmow can you share your VS version?

**Comment by nizmow** [comment]
@DHowett-MSFT sadly I no longer have access to that machine (it's a work laptop, locked safely in the office). I'm willing to be that the issue is the VS version though!

**Comment by DHowett-MSFT** [comment]
Ah, makes sense. Better safe than repro'd! I'll close this one out for now, but feel free to carp if it's still happening when we're all back to quasi-normal.
